### PR TITLE
Util to bind polymorphic subtypes

### PR DESCRIPTION
**Pull request by Randgalt**
Proposal for a JSON subtype/polymorphic binder. wdyt?

So-called polymorphic subtypes are cumbersome in Jackson. This binding utility makes it easier by:

- Moving all meta-code into the binder so that no class/record decorations/annotations are needed - de-uglifies these classes
- Automatically adds the type property to the generated JSON
- Automatically generate bindings for sealed classes/hierarchies
- Eliminates potential programmer errors by automating naming and mapping of the type field in Jackson/JSON sub-types

### CODE REVIEW

**Comment by martint** [comment]
One downside is that this forces you to use guice for constructing the json codecs. How would you do this when guice is not available?

**Comment by Randgalt** [comment]
> One downside is that this forces you to use guice for constructing the json codecs. How would you do this when guice is not available?

I could make Jackson classes public and provide a way to use them directly

**Review by jklamer — COMMENTED** [review]
Super impressive as always, especially with all the options. I would love to use this. I was wondering what you think about creating some interesting language interoperability with this?
These sealed interface types I believe are a way of creating a tagged union type (aka enum in Rust ). The defacto standard I've encountered in Rust for their JSON representation is to have the property key signify the type. Such that Programmer looks like 
```
{
    "Programmer":{
        "name":"Jack"
    }
}
```
and manager:
```
{
    "Manager":{
        "name":"Jack",
        "reports":[]
    }
}
```
The advantages of this representation being there is no need to worry about weird property conflict edge cases with the metadata field embedded into the class and we'd gain a way to map 1-1 with models in other languages. A downside being we'd lose  the ability to encode something as Employee and read as Manager (while using "Ignore unknown field property" ofc). So this wouldn't necessarily be "polymorphic" as I understand the word.

**Comment by Randgalt** [comment]
> I would love to use this. I was wondering what you think about creating some interesting language interoperability with this? These sealed interface types I believe are a way of creating a tagged union type (aka enum in Rust, and others ). The defacto standard I've encountered in Rust for their JSON representation is to have the property key signify the type. Such that Programmer looks like
> ... snip ...

Sorry it took so long to respond to this...

I don't think Jackson would support the extra level - at least I don't know how to do that in Jackson.

**Comment by Randgalt** [comment]
cc @martint and @jklamer 

I reworked this so that there is a `JsonSubType` builder that builds a new entity, `JsonSubType` that is independent of Guice. There is a new `JsonSubTypeBinder` that can bind a `JsonSubType` for Guice. Otherwise, `JsonSubType` can be used with `ObjectMapperProvider` without requiring Guice.

File: json/src/main/java/io/airlift/json/JsonSubType.java [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+        @Override
+        public B deserialize(JsonParser p, DeserializationContext ctxt)
+                throws IOException
+        {
+            TreeNode treeNode = p.readValueAsTree();
+            TreeNode propertyNode = treeNode.get(propertyName);
+
+            String propertyValue = ((propertyNode != null) && propertyNode.isValueNode()) ? ((ValueNode) propertyNode).textValue() : null;
+            if (propertyValue == null) {
+                throw new IllegalArgumentException("JSON expected to have a property named %s. Double check the addBinding() or addPermittedSubClassBindings()".formatted(propertyName));
```
Comment (electrum) [inline json/src/main/java/io/airlift/json/JsonSubType.java:219]: Add quotes around `%s`

File: json/src/main/java/io/airlift/json/JsonSubType.java [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+            }
+
+            Class<?> subClass = subClassPropertyValues.get().get(propertyValue);
+            if (subClass == null) {
+                throw new IllegalArgumentException("No binding was made for property name %s and value %s. Double check the addBinding() or addPermittedSubClassBindings().".formatted(propertyName, propertyValue));
```
Comment (electrum) [inline json/src/main/java/io/airlift/json/JsonSubType.java:224]: Same, add quotes around values

File: json/src/main/java/io/airlift/json/JsonSubType.java [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+            }
+
+            //noinspection unchecked
```
Comment (electrum) [inline json/src/main/java/io/airlift/json/JsonSubType.java:227]: I don't think this suppression works for `javac`. Probably need a helper method for this.

File: json/src/main/java/io/airlift/json/JsonSubTypeBinder.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright 2010 Proofpoint, Inc.
```
Comment (electrum) [inline json/src/main/java/io/airlift/json/JsonSubTypeBinder.java:2]: Remove, this is new code

File: json/src/main/java/io/airlift/json/JsonSubTypeBinder.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    }
+
+    public void bindJsonSubType(JsonSubType jsonSubType)
+    {
+        subTypeBinder.addBinding().toInstance(jsonSubType);
+    }
+
+    private JsonSubTypeBinder(Binder binder)
+    {
+        subTypeBinder = Multibinder.newSetBinder(binder, JsonSubType.class);
```
Comment (electrum) [inline json/src/main/java/io/airlift/json/JsonSubTypeBinder.java:37]: Nit: static import `newSetBinder`

File: json/src/test/java/io/airlift/json/Employee.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright 2010 Proofpoint, Inc.
```
Comment (electrum) [inline json/src/test/java/io/airlift/json/Employee.java:2]: Remove all of these

File: json/src/test/java/io/airlift/json/TestJsonSubType.java [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+        String item1Json = objectMapper.writeValueAsString(item1);
+        String item2Json = objectMapper.writeValueAsString(item2);
+        String containerJson = objectMapper.writeValueAsString(container);
+
+        for (int i = 0; i < 2; ++i) {
+            Part deserializedItem1 = (i == 0) ? objectMapper.readValue(item1Json, Part.class) : jsonCodec.fromJson(item1Json);
+            Part deserializedItem2 = (i == 0) ? objectMapper.readValue(item2Json, Part.class) : jsonCodec.fromJson(item2Json);
+            Part deserializedContainer = (i == 0) ? objectMapper.readValue(containerJson, Part.class) : jsonCodec.fromJson(containerJson);
+
+            assertTrue(deserializedItem1 instanceof Item);
```
Comment (electrum) [inline json/src/test/java/io/airlift/json/TestJsonSubType.java:154]: Don't use `assertTrue` with assertion expressions. Instead, use AssertJ:
```java
assertThat(deserializedItem1).isInstanceOf(Item.class);

File: json/src/main/java/io/airlift/json/JsonSubType.java [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+        {
+            this.propertyName = requireNonNull(propertyName, "propertyName is null");
+            this.subClassPropertyValues = ImmutableMap.copyOf(subClassPropertyValues);
+        }
+
+        @Override
+        public void findAndAddVirtualProperties(MapperConfig<?> config, AnnotatedClass ac, List<BeanPropertyWriter> properties)
+        {
+            String propertyValue = subClassPropertyValues.get(ac.getRawType());
+            if (propertyValue != null) {
```
Comment (electrum) [inline json/src/main/java/io/airlift/json/JsonSubType.java:171]: I'd return early here

**Review by electrum — COMMENTED** [review]
I like this.

The test coverage is good, but since it's fully automatic, we're not verifying what the JSON actually looks like (only that it round trips). We should add the expected JSON output.

Also, it would be good to have a standalone test that doesn't use Guice, to show how the APIs work and make sure Guice isn't required.

**Comment by Randgalt** [comment]
@electrum all comments addressed

**Review by electrum — APPROVED** [review]

**Pull request merged**